src/System.IO.FileSystem/tests/FileStream/WriteAsync.cs

```diff
@@ -270,6 +270,30 @@ public void WriteAsyncBufferedCompletesSynchronously()
             }
         }
 
+        [Fact]
+        public async void WriteAsyncInternalBufferOverflow()
+        {
+            // Overflow into next buffer
+            using (FileStream fs = new FileStream(GetTestFilePath(), FileMode.Create, FileAccess.Write, FileShare.None, 2))
+            {
+                // Fill existing buffer
+                await fs.WriteAsync(TestBuffer, 0, 2);
+                Assert.True(fs.Length == 2);
+
+                // Overflow into next buffer
+                await fs.WriteAsync(TestBuffer, 0, 1);
+                Assert.True(fs.Length == 3);
+
+                // Overflow bufferSize * 2
+                await fs.WriteAsync(TestBuffer, 0, 5);
+                Assert.True(fs.Length == 8);
+
+                // Overflow bufferSize * 2 with empty buffer
+                await fs.WriteAsync(TestBuffer, 0, 5);
+                Assert.True(fs.Length == 13);
+            }
+        }
```

Comment:
It'd be good to have a test that tries to issue many writes/flushes to be run concurrently. The test you have and this should also verify that Position reflects the correct value immediately after the synchronous call to the XxAsync method returns.  Note that the OS may actually choose to issue writes synchronously, and I believe it'll do so if the write would require extending the size of the file, so it'd be good to have the test verify behavior both with a zero-length file to start and with a file presized for all data needed (this coudl be done as a theory with a Boolean argument that controls whether you start by setting the length).  I believe it'll also do writes synchronously for small writes, so it'd be good to also have a test that does large writes, like 100K at a time.